src/cis2/mod.rs

```diff
@@ -0,0 +1,352 @@
+//!
+//! The type [Cis2Contract](Cis2Contract) act as a wrapper around the
+//! [Client](crate::endpoints::Client) and a contract address providing
+//! functions for querying and making transactions to smart contract.
+mod types;
+
+use crate::{common, endpoints::Client, id, types as sdk_types};
+use sdk_types::{smart_contracts, transactions, ContractAddress};
+use smart_contracts::concordium_contracts_common;
+use std::convert::{From, TryFrom};
+use thiserror::*;
+pub use types::*;
+
+/// A wrapper around the client representing a CIS2 token smart contract and
+/// provides functions for interaction.
+///
+/// Note that cloning is cheap and is therefore the intended way of sharing
+/// between multiple tasks.
+#[derive(Debug, Clone)]
+pub struct Cis2Contract<'a> {
+    client:        Client,
+    address:       ContractAddress,
+    contract_name: concordium_contracts_common::ContractName<'a>,
+}
+
+/// Error which can occur when submitting a transaction such as `tranfer` and
+/// `updateOperator` to a CIS2 smart contract.
+#[derive(Debug, Error)]
+pub enum Cis2TransactionError {
+    /// The smart contract receive name is invalid.
+    #[error("Invalid receive name: {0}")]
+    InvalidContractName(concordium_contracts_common::NewReceiveNameError),
+
+    /// A general RPC error occured.
+    #[error("RPC error: {0}")]
+    RPCError(#[from] crate::endpoints::RPCError),
+}
+
+impl From<concordium_contracts_common::NewReceiveNameError> for Cis2TransactionError {
+    fn from(err: concordium_contracts_common::NewReceiveNameError) -> Self {
+        Self::InvalidContractName(err)
+    }
+}
+
+/// Error which can occur when invoking a query such as `balanceOf` and
+/// `operatorOf` or `tokenMetadata` to a CIS2 smart contract.
+#[derive(Debug, Error)]
+pub enum Cis2QueryError {
+    /// The smart contract receive name is invalid.
+    #[error("Invalid receive name: {0}")]
+    InvalidContractName(concordium_contracts_common::NewReceiveNameError),
+
+    /// A general RPC error occured.
+    #[error("RPC error: {0}")]
+    RPCError(#[from] crate::endpoints::RPCError),
+
+    /// The returned bytes from invoking the smart contract could not be parsed.
+    #[error("Failed parsing the response")]
+    ResponseParseError(concordium_contracts_common::ParseError),
+
+    /// The node rejected the invocation.
+    #[error("Rejected by the node")]
+    NodeRejected(sdk_types::RejectReason),
+}
+
+impl From<concordium_contracts_common::NewReceiveNameError> for Cis2QueryError {
+    fn from(err: concordium_contracts_common::NewReceiveNameError) -> Self {
+        Self::InvalidContractName(err)
+    }
+}
+
+impl From<concordium_contracts_common::ParseError> for Cis2QueryError {
+    fn from(err: concordium_contracts_common::ParseError) -> Self { Self::ResponseParseError(err) }
+}
+
+impl From<sdk_types::RejectReason> for Cis2QueryError {
+    fn from(err: sdk_types::RejectReason) -> Self { Self::NodeRejected(err) }
+}
```

Comment:
Recommend adding a comment for future maintainers as to why we are implementing this manually.